### PR TITLE
Replace CountedPtr by std::shared_ptr in lattices

### PR DESCRIPTION
**Pull request by gervandiepen**
This also contains a change in casa/Utilities/COWPtr which is used by lattices.
The deleteIt option in COWPtr needs to be maintained, because it is used by lattices to avoid copying Array objects.

### CODE REVIEW

File: lattices/LEL/test/tLEL.cc [hunk 1]
```diff
@@ -316,7 +316,7 @@ int main (int argc, const char* argv[])
 //
    cout << endl << "LELUnary<Float>" << endl;
   {
-    CountedPtr<LELInterface<Float> > pExpr = new LELLattice<Float>(bF);
+    std::shared_ptr<LELInterface<Float> > pExpr(new LELLattice<Float>(bF));
```
Comment (mnijhuis-tos) [inline lattices/LEL/test/tLEL.cc:319]: `std::make_shared` avoids the duplicate `LELInterface<Float>` type here. There are many other similar occurences in this file and other files below, that can be shortened this way.
```suggestion
   auto pExpr = std::make_shared<LELLattice<Float>>(bF);
```

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]
My main concern is the newly added dCOWptr.out file. My other comments are just further improvement suggestions.

File: images/Images/ImageProxy.h [hunk 2]
```diff
@@ -464,13 +464,13 @@ namespace casacore {
...
-    ImageInterface<Double>*   itsImageDouble;
-    ImageInterface<Complex>*  itsImageComplex;
-    ImageInterface<DComplex>* itsImageDComplex;
-    const CoordinateSystem*   itsCoordSys;
-    ImageAttrHandler*         itsAttrHandler;
+    std::shared_ptr<LatticeBase>   itsLattice;
+    ImageInterface<Float>*         itsImageFloat;
+    ImageInterface<Double>*        itsImageDouble;
+    ImageInterface<Complex>*       itsImageComplex;
+    ImageInterface<DComplex>*      itsImageDComplex;
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply images/Images/ImageProxy.h:471]: That's not needed. As said in the comments above shared_ptr<LatticeBase> does the reference counting. Therefore a shared_ptr for the ImageInterface cannot be used.
I could remove shared_ptr<LatticeBase> and have 4 shared_ptr objects of which 3 will be empty, but see no advantage in it.

File: lattices/LEL/test/tLEL.cc (hunk 1 above)
Comment (gervandiepen) [reply lattices/LEL/test/tLEL.cc:319]: Now using make_shared in lattices test programs and in a few other source files.

File: casa/Utilities/COWPtr.h [hunk 3]
```diff
@@ -261,47 +266,35 @@ template <class T> class COWPtr
...
-// Make our own default pointer - deleteIt==True by default, const_p==False
+//# Make our own default pointer - deleteIt==True by default, const_p==False
 template <class T> inline COWPtr<T>::COWPtr()
-:obj_p(static_cast<T *>(0), True), const_p(False)
-{
-  // does nothing
-} 
+  : obj_p   (nullptr, Deleter(True)),
+    const_p (False)
+{}
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply casa/Utilities/COWPtr.h:279]: Changed.

File: casa/Utilities/COWPtr.h [hunk 4]
```diff
@@ -261,47 +266,37 @@ template <class T> class COWPtr
...
-// copy ctor with reference semantics
+//# copy ctor with reference semantics
 template <class T> inline COWPtr<T>::COWPtr(const COWPtr<T> &other)
-: obj_p(other.obj_p), const_p(other.const_p)
-{
-  // does nothing
-}
+  : obj_p   (other.obj_p),
+    const_p (other.const_p)
+{}
```
Comment (mnijhuis-tos) [inline casa/Utilities/COWPtr.h:287]:
```suggestion
{
  // does nothing
}
```

File: casa/Utilities/COWPtr.tcc [hunk 5]
```diff
@@ -31,30 +31,31 @@
 namespace casacore { //# NAMESPACE CASACORE - BEGIN
 
 template <class T> COWPtr<T>::COWPtr(T *obj, Bool deleteIt, Bool readOnly)
-: obj_p(obj, deleteIt), const_p(readOnly)
-{
-  // does nothing
-}
+  : obj_p   (obj, Deleter(deleteIt)),
+    const_p (readOnly)
+{}
```
Comment (mnijhuis-tos) [inline casa/Utilities/COWPtr.tcc:36]:
```suggestion
{
  // does nothing
}
```

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: images/Images/FITSImage.cc [hunk 6]
```diff
@@ -172,7 +167,6 @@ FITSImage& FITSImage::operator=(const FITSImage& other)
  
 FITSImage::~FITSImage()
 {
-   delete pPixelMask_p;
 }
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply images/Images/FITSImage.cc]: Declared as default in FITSImage.h and removed from .cc file

File: casa/Utilities/COWPtr.tcc (hunk 5 above)
Comment (gervandiepen) [reply casa/Utilities/COWPtr.tcc]: done

File: casa/Utilities/COWPtr.h (hunk 4 above)
Comment (gervandiepen) [reply casa/Utilities/COWPtr.h]: done

**Pull request merged**